handler/src/main/java/io/netty/handler/ssl/OpenSslServerContext.java

```diff
@@ -349,74 +346,14 @@ private OpenSslServerContext(
-                                    new OpenSslExtendedKeyMaterialManager(
-                                            (X509ExtendedKeyManager) keyManager, keyPassword) :
-                                    new OpenSslKeyMaterialManager(keyManager, keyPassword);
-                        } else {
-                            keyMaterialManager = null;
-                        }
-                    }
-                } catch (Exception e) {
-                    throw new SSLException("failed to set certificate and key", e);
-                }
-                try {
-                    if (trustCertCollection != null) {
-                        trustManagerFactory = buildTrustManagerFactory(trustCertCollection, trustManagerFactory);
-                    } else if (trustManagerFactory == null) {
-                        // Mimic the way SSLContext.getInstance(KeyManager[], null, null) works
-                        trustManagerFactory = TrustManagerFactory.getInstance(
-                                TrustManagerFactory.getDefaultAlgorithm());
-                        trustManagerFactory.init((KeyStore) null);
-                    }
-
-                    final X509TrustManager manager = chooseTrustManager(trustManagerFactory.getTrustManagers());
-
-                    // IMPORTANT: The callbacks set for verification must be static to prevent memory leak as
-                    //            otherwise the context can never be collected. This is because the JNI code holds
-                    //            a global reference to the callbacks.
-                    //
-                    //            See https://github.com/netty/netty/issues/5372
-
-                    // Use this to prevent an error when running on java < 7
-                    if (useExtendedTrustManager(manager)) {
-                        SSLContext.setCertVerifyCallback(ctx,
-                                new ExtendedTrustManagerVerifyCallback(engineMap, (X509ExtendedTrustManager) manager));
-                    } else {
-                        SSLContext.setCertVerifyCallback(ctx, new TrustManagerVerifyCallback(engineMap, manager));
-                    }
-                } catch (Exception e) {
-                    throw new SSLException("unable to setup trustmanager", e);
-                }
-            }
-            sessionContext = new OpenSslServerSessionContext(this);
-            sessionContext.setSessionIdContext(ID);
+            ServerContext context = newSessionContext(this, ctx, engineMap, trustCertCollection, trustManagerFactory,
+                                                      keyCertChain, key, keyPassword, keyManagerFactory);
+            sessionContext = context.sessionContext;
+            keyMaterialManager = context.keyMaterialManager;
             success = true;
         } finally {
             if (!success) {
-                destroy();
+                release();
```

Comment:
the existing behavior propagates the exception to the caller. In the case of the finalizer there is no "caller" which has visibility on the exception.